### PR TITLE
Start exposing a new COM-lite API

### PR DESCRIPTION
**Pull request by tfoleyNV**
This change is mostly about exposing a new API to the Slang compiler that allows more fine-grained control over the compilation flow. The basic concepts in the new API are:

* An `IGlobalSession` is the granularity at which we load/parse the Slang stdlib, and therefore gives applications a way to amortize startup cost for the library across multiple compiles. This is a concept that might be able to go away in a future version of Slang.

* An `ISession` owns all the code that gets loaded/compiled/generated. Any `import`ed modules are shared across everything in a session (we don't re-parse/-check the code when we see another `import` for the same module). Any generic- or interface-based code in the session can be specialized using types from the same session (but not necessarily across sessions).

* An `IModule` is the unit of code loading and scoping. It doesn't expose any API in this change, but would be the right scope for looking up types or entry points by name.

* An `IProgram` is a "linked" combination of modules and entry points from which code can be generated and reflection information queried.

This change re-uses the existing reflection API types, rather than introduce a new API that duplicates that functionality. That will probably change in a future revision.

There are two major pieces of functionality added here that aren't related to the new API:

* We now have an API concept of "entry point groups" which are one or more entry points that are intended to be used together so that they need to have non-overlapping parameters. For now this is being used to handle "hit groups" and local root signatures for ray tracing, but I'm not sure this is a concept we will keep in the long run.

* We have a very special-case (client-application-specific) flag that ascribes special meaning to the `shared` keyword, so that it can be attached to global parameters to indicate that they are actually to be part of the local root signature rather than the global one for DXR.

None of the API design (including naming) here is finalized; the only reason to check in the changes at this point to avoid having a long-running branch that leads to merge pain. Clients should *not* try to depend on the new API just yet, since it is still a work in progress.

### CODE REVIEW

File: slang.h [hunk 1]
```diff
@@ -2543,8 +2597,329 @@ namespace slang
...
+        Kind kind;
+        union
+        {
+            TypeReflection* type;
+        };
+    };
+
+        /** Description of a program to be created.
+        */
+    struct ProgramDesc
```
Comment (jsmall-nvidia) [inline slang.h]: This may be best way to do this. But if there is a base interface (perhaps having no new methods) that both inherited from, then perhaps it might be cleaner to just use an array of pointers to that type, and use dynamic casting internally.

File: slang.h [hunk 2]
```diff
@@ -2543,8 +2597,329 @@ namespace slang
...
+    {
+        struct Item
```
Comment (jsmall-nvidia) [inline slang.h]: Just note that there is no binary backwards compatibility in Desc'y types.

File: slang.h [hunk 3]
```diff
@@ -2543,8 +2597,329 @@ namespace slang
...
+        };
+
+        Item const* items;
+        SlangInt    itemCount;
+    };
+
+
+        /** A program comprises zero or more modules, entry points, etc. that have been linked together.
+        */
+    struct IProgram : public ISlangUnknown
```
Comment (jsmall-nvidia) [inline slang.h]: IProgram now seems more reasonable - in the context of being used similarly in other APIs.

File: slang.h [hunk 4]
```diff
@@ -2543,8 +2597,329 @@ namespace slang
...
+            SlangInt    targetIndex = 0,
+            IBlob**     outDiagnostics = nullptr) = 0;
+
+            /** Get the compiled code for the entry point at `entryPointIndex` for the chosen `targetIndex`
+
+            If code has not already been generated for the given entry point and target,
+            then a compilation error may be detected, in which case `outDiagnostics`
+            (if non-null) will be filled in with a blob of messages diagnosing the error.
+            */
+        virtual SLANG_NO_THROW SlangResult SLANG_MCALL getEntryPointCode(
```
Comment (jsmall-nvidia) [inline slang.h]: So this is lazily compiling? If so it might be better to have verb other than 'get' to make clear it could do quite a bit of work.

File: source/slang/slang-compiler.h [hunk 5]
```diff
@@ -1277,9 +1421,24 @@ namespace Slang
     struct TypeCheckingCache;
     //
 
-    class Session
+    class Session : public RefObject, public slang::IGlobalSession
```
Comment (jsmall-nvidia) [inline source/slang/slang-compiler.h]: Do we want to rename Session -> GlobalSession?

File: source/slang/slang-compiler.h [hunk 6]
```diff
@@ -1421,6 +1580,76 @@ namespace Slang
         RefPtr<Linkage> m_builtinLinkage;
     };
 
+
+//
+// The following functions are utilties to convert between
```
Comment (jsmall-nvidia) [inline source/slang/slang-compiler.h]: I'm not impartial - but this seems good to me :)

File: source/slang/slang-reflection.cpp [hunk 7]
```diff
@@ -86,6 +86,15 @@ static inline SlangReflectionEntryPoint* convert(EntryPointLayout* entryPoint)
     return (SlangReflectionEntryPoint*) entryPoint;
 }
 
+static inline EntryPointGroupLayout* convert(SlangEntryPointGroupLayout* entryPointGroup)
```
Comment (jsmall-nvidia) [inline source/slang/slang-reflection.cpp]: Not an asInternal?

File: source/slang/slang.cpp [hunk 8]
```diff
@@ -91,6 +97,55 @@ Session::Session()
...
+    if(guid == IID_ISlangUnknown || guid == IID_IGlobalSession)
+        return asExternal(this);
+    return nullptr;
+}
+
+SLANG_NO_THROW SlangResult SLANG_MCALL Session::createSession(
+    slang::SessionDesc const&  desc,
+    slang::ISession**          outSession)
+{
+    auto linkage = new Linkage(this);
```
Comment (jsmall-nvidia) [inline source/slang/slang.cpp]: Might have been better to use a smart pointer here, and then detach at the very end. Means returns/throw inbetween don't leak.

File: source/slang/slang.cpp [hunk 9]
```diff
@@ -330,9 +385,183 @@ Linkage::Linkage(Session* session)
...
+                program->addReferencedModule(module);
+            }
+            break;
+
+        default:
+            return SLANG_E_INVALID_ARG;
+        }
+    }
+
+    *outProgram = ComPtr<slang::IProgram>(asExternal(program)).detach();
```
Comment (jsmall-nvidia) [inline source/slang/slang.cpp]: Can't we avoid the extra com AddRef? Using ```*outProgram = asExternal(program.detach());```

File: source/slang/slang.cpp [hunk 10]
```diff
@@ -785,13 +1036,15 @@ SlangResult FrontEndCompileRequest::executeActionsInner()
...
         return SLANG_FAIL;
 
-    if ((compileFlags & SLANG_COMPILE_FLAG_NO_CODEGEN) == 0)
-    {
-        // Generate initial IR for all the translation
-        // units, if we are in a mode where IR is called for.
-        generateIR();
-    }
-
+    // We always generate IR for all the translation units.
```
Comment (jsmall-nvidia) [inline source/slang/slang.cpp]: What happened to NO_CODEGEN? That I know people are using it to find dependencies.

File: source/slang/slang-parameter-binding.cpp [hunk 11]
```diff
@@ -585,7 +588,7 @@ static bool findLayoutArg(
         if( modifier )
         {
             *outVal = (UInt) strtoull(String(modifier->valToken.Content).getBuffer(), nullptr, 10);
-            return true;
+return true;
```
Comment (jsmall-nvidia) [inline source/slang/slang-parameter-binding.cpp]: tabbing problem?

File: source/slang/slang-parameter-binding.cpp [hunk 12]
```diff
@@ -2510,17 +2568,75 @@ RefPtr<ProgramLayout> generateParameterBindings(
...
+    // After we have allocated registers/bindings to everything
+    // in the global scope we will process the parameters
+    // of the entry points.
+    //
+    // Note: at the moment we are laying out *all* information related to global-scope
+    // parameters (including pending data from interface-type parameters) before
+    // anything pertaining to entry points. This is a crucial design choice to
+    // get right, and we might want to revisit it based on experience.
+
+    // In some cases, a user will want o ensure that all the
```
Comment (jsmall-nvidia) [inline source/slang/slang-parameter-binding.cpp]: to

Comment (jsmall-nvidia) [reply source/slang/slang-parameter-binding.cpp]: overallping

File: slang.h [hunk 13]
```diff
@@ -2543,8 +2597,329 @@ namespace slang
...
+            SlangInt                    specializationArgCount,
+            SpecializationArg const*    specializationArgs,
+            IProgram**                  outSpecializedProgram,
+            ISlangBlob**                outDiagnostics = nullptr) = 0;
+
+            /** Specialize a type based on type arguments.
+            */
+        virtual SLANG_NO_THROW TypeReflection* SLANG_MCALL specializeType(
+            TypeReflection*             type,
+            SlangInt                    specializationArgCount,
```
[the remark this answers is not in the document]
Comment (tfoleyNV) [reply slang.h]: I think I agree that `Foo* foos, int fooCount` is the more reasonable alternative.

File: slang.h [hunk 14]
```diff
@@ -2543,8 +2597,329 @@ namespace slang
...
+            SlangInt        targetIndex = 0,
+            LayoutRules     rules = LayoutRules::Default,
+            ISlangBlob**    outDiagnostics = nullptr) = 0;
+
+            /** Create a request to load/compile front-end code.
+            */
+        virtual SLANG_NO_THROW SlangResult SLANG_MCALL createCompileRequest(
+            SlangCompileRequest**   outCompileRequest) = 0;
+
+#if 0
```
[the remark this answers is not in the document]
Comment (tfoleyNV) [reply slang.h]: Yes, that is cruft left over from the world before I added all that state to the `Desc`.

One note is that there is nothing about the `ISession` implementation that precludes adding a target after creation, so if it ever turns out to be limiting that we require all targets to be registered up front, we could change that.

File: slang.h (hunk 1 above)
Comment (tfoleyNV) [reply slang.h]: That's a cleanup I think will happen automatically if I do the big refactor I'm thinking of. If I introduce a base interface for (lacking a better term) "component types," then a component type would be the base interface, and all of the current "module," "entry point," and "program" concepts would fit into it. Also the new "entry point group" concept would probably not need to be a separate thing. The main catch would be that the new concept probably ends up needing its own reflection API...

File: slang.h (hunk 2 above)
Comment (tfoleyNV) [reply slang.h]: Yes, I haven't added the hooks for backwards compatibility here yet, and there is *no* intention to support the API in this PR as-is in the long run (this is a milestone toward the first COM-lite API that will provide backwards compatibility, but it is not yet that point).

I'm still on the fence about whether to do the `sizeof` trick or Vulkan-style self-identifying structure types with support for extensions. My gut wants to do the latter (so that seldom-used features don't clutter up the structures for ordinary users), but my brain says that is probably over-engineering things.

File: slang.h (hunk 3 above)
Comment (tfoleyNV) [reply slang.h]: Okay. I don't think it has actually *changed* since the earlier pitches, so maybe it is just more clear in context of the API without the confusion of "linkage."

File: slang.h (hunk 4 above)
Comment (tfoleyNV) [reply slang.h]: Yes, the idea is that you can in principle use the Slang API as your "shader cache" layer, and if/when we support loading from binary container files that could include pre-computed kernels, a naive app that does aggressive on-demand compilation could Just Work with our caching layer.

I agree that the naming is maybe under-selling the amount of work. At the implementation level this is a `getOrCreate...` instead of just `get...` which is more accurate, but clumsy.

I don't want the verb to imply that code generation happens every time.

File: source/slang/slang-compiler.h (hunk 5 above)
Comment (tfoleyNV) [reply source/slang/slang-compiler.h]: Yeah, that would be helpful, but I think that needs to be a distinct refactoring/renaming pass because it will change so many source files that it could make the diffs confusing. If we want to do that rename as a separate PR *first*, I could do that.

File: source/slang/slang-reflection.cpp (hunk 7 above)
Comment (tfoleyNV) [reply source/slang/slang-reflection.cpp]: The old functions that support the C-style API have been left as-is with their original naming.

File: source/slang/slang.cpp (hunk 8 above)
Comment (tfoleyNV) [reply source/slang/slang.cpp]: Oh yeah, that looks like an oversight.

File: source/slang/slang.cpp (hunk 9 above)
Comment (tfoleyNV) [reply source/slang/slang.cpp]: Probably?

File: source/slang/slang.cpp (hunk 10 above)
Comment (tfoleyNV) [reply source/slang/slang.cpp]: The flags in the compiler were a bit of a mess, because there were two similarly-named options that did two nearly-but-not-quite-the-same things:

1. There was an option to suppress final kernel code generation (equivalent to not having a target, really), but that *didn't* suppress IR code generation.

2. There was an option to suppress *all* code generation, including IR code generation.

I believe what I've done here is eliminate option (2) and kept option (1), so the main cost for somebody who just wants to resolve dependencies is that we end up generating IR from the type-checked AST, instead of skipping that step. I could try to restore the old behavior, but we'd need two distinct flags/options for this, because Falcor needs (1), and that is the most sane default for the average user.

File: source/slang/slang-parameter-binding.cpp (hunk 11 above)
Comment (tfoleyNV) [reply source/slang/slang-parameter-binding.cpp]: Looks like it. My editor seems to have been doing some weird things where it is changing indentation on lines unrelated to my edits.
